dev/DynamicDependency/PackageGraphManager.cpp

```diff
@@ -0,0 +1,241 @@
+            LONG rc{ getCurrentPackageInfo2(flags, packagePathType, &staticPackageGraphBufferLength, nullptr, nullptr) };
+            if (rc == APPMODEL_ERROR_NO_PACKAGE)
+            {
+                // The process has no static package info
+                // Nothing to do
+            }
+            else if (rc == ERROR_INSUFFICIENT_BUFFER)
+            {
+                // We've got 1+ match! Get the data
+                staticPackageGraphBuffer = wil::make_unique_cotaskmem<BYTE[]>(staticPackageGraphBufferLength);
+                RETURN_IF_WIN32_ERROR(getCurrentPackageInfo2(flags, packagePathType, &staticPackageGraphBufferLength, staticPackageGraphBuffer.get(), &staticPackagesCount));
+                staticPackageInfo = reinterpret_cast<PACKAGE_INFO*>(staticPackageGraphBuffer.get());
+            }
+            else
+            {
+                RETURN_WIN32(rc);
+            }
+        }
+    }
+
+    // Update the total 'count' (if any)
+    const auto totalPackagesCount{ staticPackagesCount + dynamicPackagesCount };
+    if (count)
+    {
+        *count = totalPackagesCount;
+    }
+
+    // Do we have anything to report?
+    if (totalPackagesCount == 0)
+    {
+        // No matches!
+        *bufferLength = 0;
+        return S_OK;
+    }
+
+    // Compute the buffer length used/needed and fill buffer (if we can)
+    auto bufferNeeded{ SerializePackageInfoToBuffer(flags, packagePathType, *bufferLength, buffer, matchingPackageInfo, dynamicPackagesCount, staticPackageInfo, staticPackagesCount) };
+    const auto isInsufficientBuffer{ *bufferLength < bufferNeeded };
+    *bufferLength = bufferNeeded;
+    RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER), isInsufficientBuffer);
+
+    return S_OK;
+}
+CATCH_RETURN();
+
+UINT32 MddCore::PackageGraphManager::SerializePackageInfoToBuffer(
+    const UINT32 flags,
+    const PackagePathType packagePathType,
+    const UINT32 bufferLength,
+    BYTE* buffer,
+    const std::vector<const MddCore::PackageGraphNode*>& matchingPackageInfo,
+    const UINT32 dynamicPackagesCount,
+    const PACKAGE_INFO* staticPackageInfo,
+    const UINT32 staticPackagesCount)
+{
+    // Setup our initial pointers to write the PACKAGE_INFO[] and the variable length data
+    const auto totalPackagesCount{ dynamicPackagesCount + staticPackagesCount };
+    UINT32 bufferNeeded{ sizeof(PACKAGE_INFO) * totalPackagesCount };
+    auto toPackageInfo{ reinterpret_cast<PACKAGE_INFO*>(buffer) };
+
+    // Let's do it...
+    for (auto& matchingPackageGraphNode : matchingPackageInfo)
+    {
+        UINT32 fromPackagesCount{};
+        const void* fromPackagesBuffer{};
+        wil::unique_cotaskmem_ptr<BYTE[]> dynamicPackagesBuffer;
+
+        if (matchingPackageGraphNode)
+        {
+            // It's dynamic package data
+            fromPackagesCount = std::move(matchingPackageGraphNode->GetMatchingPackages(flags, packagePathType, dynamicPackagesBuffer));
```

Comment:
Scalars don't need "move" semantics, simple assignment is fine.

Reply:
+1